### PR TITLE
Coin market data recovery scheme was added

### PR DESCRIPTION
**Pull request by waqarRns**

### CODE REVIEW

File: src/Stoa.ts [hunk 1]
```diff
@@ -1658,6 +1661,43 @@ class Stoa extends WebService
             return resolve({ page, pageSize });
         });
     }
+    /**
+     * 
+     */
+    private async getBoaPriceChart(req: express.Request, res: express.Response)
```
Comment (TrustHenry) [inline src/Stoa.ts]: Please add a document.

File: src/modules/service/CoinMaketService.ts [hunk 2]
```diff
@@ -28,34 +31,90 @@ export class CoinMarketService {
...
-                });
+            this.status = false;
+            cron.schedule(`*/${time} * * * * *`, async () => {
+                await this.scheduler(stoaInstance);
+                resolve(true);
             })
+
+        });
+    }
+    public scheduler(stoaInstance: Stoa): Promise<boolean> {
```
Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts:67]: Please add a document.

Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts]:
```suggestion
    }
    
    public scheduler(stoaInstance: Stoa): Promise<boolean> {
```

File: src/modules/service/CoinMaketService.ts [hunk 3]
```diff
@@ -28,34 +31,90 @@ export class CoinMarketService {
      */
     private job: cron.ScheduledTask | null = null;
 
+    /**
+     *  
+     */
+    private status: boolean;
```
Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts]: Please add a document.

File: src/modules/service/CoinMaketService.ts [hunk 4]
```diff
@@ -28,34 +31,90 @@ export class CoinMarketService {
...
+                marketCap.forEach(async (element: IMarketCap) => {
+                    let time = await Time.msToTime(element.last_updated_at);
+                    element.last_updated_at = time.seconds;
+                    await stoaInstance.putCoinMarketStats(element);
+                });
+                resolve(true)
+            }
+            catch (err) {
+                logger.error(`Error: while 24 hour coin maket recovery`);
+                reject(`Error: while 24 hour coin maket recovery`);
```
Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts]:
```suggestion
                logger.error(`Failed to 24-hour coin market data recovery: ${err}`);
                reject(`Failed to 24-hour coin market data recovery`);
```

File: src/modules/service/CoinMaketService.ts [hunk 5]
```diff
@@ -64,18 +123,31 @@ export class CoinMarketService {
...
+                    return resolve(true)
                 }
-
-            }).catch((err) => {
-                logger.error("Failed to recover data: " + err);
-            })
+            }
+            catch (err) {
+                logger.error(`Error: while coin maket recovery`)
+                reject(`Error: while coin maket recovery`)
```
Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts]:
```suggestion
                logger.error(`Failed to coin market data recovery: ${err}`)
                reject(`Failed to coin market data recovery`)
```

File: src/modules/service/CoinMaketService.ts [hunk 6]
```diff
@@ -28,34 +31,90 @@ export class CoinMarketService {
...
+                    }
+                }).catch(async (err: any) => {
+                    await this.recover24hourData(stoaInstance);
+                    this.status = false;
+                    resolve(true)
+                });
+            }
+        });
+    }
+    /**
```
Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts]:
```suggestion
    }
    
    /**
```

File: src/modules/service/CoinMaketService.ts [hunk 7]
```diff
@@ -28,34 +31,90 @@ export class CoinMarketService {
...
+     */
+    public recover24hourData(stoaInstance: Stoa): Promise<boolean> {
+        return new Promise<boolean>(async (resolve, reject) => {
+            try {
+                let to = await Time.msToTime(Date.now());
+                let from = await JSBI.subtract(JSBI.BigInt(to.seconds), JSBI.BigInt(60 * 60 * 24));
+                let dt = new Date(to.seconds * 1000);
+                let df = new Date(Number(from.toString()) * 1000)
+                logger.info(`Recovering 24 hour coinmarket cap from: ${df} to: ${dt}`)
+                let marketCap = await this.coinMarketClient.recover(Number(from.toString()), to.seconds);
```
Comment (TrustHenry) [inline src/modules/service/CoinMaketService.ts:113]: err: TypeError: this.coinMarketClient.recover is not a function
Errors are occurring from time to time. (TestCoinGecko)

**Review by TrustHenry — CHANGES_REQUESTED** [review]

**Comment by TrustHenry** [comment]
Can you rebase the PR to Stoa/v0.x.x branch?

File: src/modules/service/CoinMaketService.ts (hunk 7 above)
Comment (waqarRns) [reply src/modules/service/CoinMaketService.ts:113]: resolved

**Pull request closed without merging**